Upwind flux for the CurvedScalarWave system

prayush opened the pull request:
## Proposed changes

Adding the upwind flux for the scalar-wave system in curved background spacetime. 


### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

prayush committed:
Add upwind flux for the CurvedScalarWave system

prayush commented:
**Note for reviewers**: In #1760 I have already renamed char fields to `v`. The older version of #1760 (which uses `u` for char field vars) is included in here as the first commit. Once #1760 is merged, I will rebase this PR onto develop, and rename `u` to `v` along with it. Please review this PR keeping this caveat in mind :slightly_smiling_face: !

**Update (01/16/2020)**: Char variables have been renamed from `u` to `v` here throughout. The above note is not valid now. #1760 has already been merged

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -92,4 +86,203 @@ struct ComputeDuDt {
+ * theorem). This gives:
+ *
+ * \f{align*}
+ * F(\psi) &= -(1 + \gamma_1) N^k n_k \psi \\
```
Please use our standard notation, such as alpha for lapse, etc.

prayush replied:
changed them everywhere

geoffrey4444 reviewed (changes requested):
not a full review, but noticed notation in help text needs updating

prayush commented:
Hi @sxs-collaboration/spectre-core-devs - An unrelated unit test is failing for Travis here, 

```
0.135 s: Unit.Numerical.LinearOperators.Filter
=====================================
test cases:     1 |     0 passed | 1 failed
assertions: 12096 | 12093 passed | 3 failed
```

with

```
FAILED:
CHECK( a == appx(b) )
with expansion:
  -680.8670826749 == Approx( -680.867082675 ).
```

If this is a precision issue, it should fail for other PRs as well at random. Have you guys seen such failures before?

prayush commented:
> not a full review, but noticed notation in help text needs updating

Hi @geoffrey4444 - I have updated notation in help text everywhere in this PR, and rebased it onto the latest version of #1760

nilsdeppe commented:
> Hi @sxs-collaboration/spectre-core-devs - An unrelated unit test is failing for Travis,
> 
> ```
> 0.135 s: Unit.Numerical.LinearOperators.Filter
> =====================================
> test cases:     1 |     0 passed | 1 failed
> assertions: 12096 | 12093 passed | 3 failed
> ```
> 
> with
> 
> ```
> FAILED:
> CHECK( a == appx(b) )
> with expansion:
>   -680.8670826749 == Approx( -680.867082675 ).
> ```
> 
> If this is a precision issue, it should fail for other PRs as well at random. Have you guys seen such failures before?

Yes, I have it on my todo list to fix this :)

prayush commented:
> > If this is a precision issue, it should fail for other PRs as well at random. Have you guys seen such failures before?
> 
> Yes, I have it on my todo list to fix this :)

Aha, the same just happened in #1752 (this [test](https://github.com/sxs-collaboration/spectre/pull/1752/checks?check_run_id=393728866))

prayush commented:
ping @geoffrey4444

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ * of the element, and \f$\hat{U}^{\rm ext}= S^{-1}U^{\rm ext}\f$ are the
+ * same computed from evolved variables taken from the element exterior, i.e.
+ * the neighboring element.
+ * Further, the sign of characteristic speeds indicates the direction of
```
The sign of characteristic speeds... (no Further,)

prayush replied:
changed

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ * propagation of the corresponding characteristic field with respect to the
+ * interface normal that the field has been
+ * computed along (with negative speeds indicating incoming characteristics,
+ * and vice-versa). Therefore, \f$\Lambda^+\f$ contains characterstic speeds
```
and with positive speeds indicating outgoing characteristics

prayush replied:
reworded

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ * computed along (with negative speeds indicating incoming characteristics,
+ * and vice-versa). Therefore, \f$\Lambda^+\f$ contains characterstic speeds
+ * for variables that are outgoing from the element at the interface in
+ * question, and vice-versa for \f$\Lambda^-\f$. An ambiguity naturally arises
```
remove "in question", and change `and vice versa for \f$\Lambda^-\f$` to `and \f$\Lambda^-\f$ contains characteristic speeds for variables that are incoming to the element at the interface.`

prayush replied:
changed

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ * here also we compute both and use their average, i.e.
+ * \f$\lambda^{\rm avg} = (\lambda^{\rm int} + \lambda^{\rm ext})/2\f$, to
+ * populate \f$\Lambda^\pm\f$.
+ *
```
I think this treatment is fine as long as there are no shocks, as is the case for GH. Is that the case for scalar wave on a curved background?

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ *
+ * This function computes the upwind flux as follows:
+ *  -# Computes characteristic variables and speeds using evolved variables from
+ *     both, the element interior and exterior.
```
no comma after both

prayush replied:
removed

geoffrey4444 commented on src/Evolution/Systems/CurvedScalarWave/Equations.hpp:
```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+ *
+ *
+ * This function computes the upwind flux as follows:
+ *  -# Computes characteristic variables and speeds using evolved variables from
```
`Computes internal and external characteristic variables and speeds`

prayush replied:
changed

geoffrey4444 reviewed (changes requested):
Looks reasonable to me. I have some suggested changes on the documentation.

This PR copies and pastes a lot of code, with only slight differences, from the GH upwind flux. I have no significant suggested changes to the code itself, since all of this is the same as the upwind flux code already reviewed for GH. 

I kind of wish there were a way to make an upwind flux once, instead of (almost) duplicating code. But I don't see a straightforward way to do that, so I'm ok with this as-is. If someone has a nice way to refactor this without a lot of work, to minimize code duplication, I'd be all for it, though.

prayush commented:
> This PR copies and pastes a lot of code, with only slight differences, from the GH upwind flux. I have no significant suggested changes to the code itself, since all of this is the same as the upwind flux code already reviewed for GH.
> 
> I kind of wish there were a way to make an upwind flux once, instead of (almost) duplicating code. But I don't see a straightforward way to do that, so I'm ok with this as-is. If someone has a nice way to refactor this without a lot of work, to minimize code duplication, I'd be all for it, though.

I couldn't agree more, and had wished that already when trying to cobble together the curved SW executable. Its quite counterproductive to implement similar formulae in two different but related evol systems. For instance, The helper function `weight_char_field` can be shared as is with GH, `weight_char_fields` can be also shared once char field tags are renamed from `U*` to `V*` in GH, but I have no idea how `ComputeNormalDotFluxes` or `UpwindFlux` could easily share code with GH as the tensorial nature of evolved fields is explicitly invoked in them..

nilsvu commented:
@prayush the issue of code duplication is related to my comment on https://github.com/sxs-collaboration/spectre/pull/1852. Some template-metaprogramming can help here. Sharing code also means you have to write fewer tests and there’s less code to maintain. If you like we can meet in a hangout some time next week and see what can be done :) just ping me on Slack if you want to do that.

prayush commented:
> @prayush the issue of code duplication is related to my comment on #1852. Some template-metaprogramming can help here. Sharing code also means you have to write fewer tests and there’s less code to maintain. If you like we can meet in a hangout some time next week and see what can be done :) just ping me on Slack if you want to do that.

Hi @nilsleiffischer - apologies for not getting back to this earlier. I agree that we should have as little code duplication as possible, but as @geoffrey4444 noted [above](https://github.com/sxs-collaboration/spectre/pull/1770#pullrequestreview-360423711) I do not see an easy way to share the upwind flux code across systems. Can we discuss possibilities? I am happy to jump into a hangout sometime, or continue the discussion here itself.

nilsdeppe commented:
@geoffrey4444 this is still waiting on you...

nilsdeppe commented:
@nilsleiffischer there was a reply to your comment, I'm not sure if there was any more discussion, but if so it would be great to summarize here.

nilsdeppe commented:
I'm not worried about the code duplication. The boundary term I've derived that handles upwinding more cleanly will be easy to generalize for handling any system that has 4 char speeds with correspanding char fields in the same form as the scalar wave systems (includes SW, CSW, and GH). Also, going forward we won't need a `ComputeNormalDotFlux` for nonconservative systems.
